File: leetcode-solution/leetcode74/Solution.java
public class Solution {
    /**
     * @param matrix, a list of lists of integers
     * @param target, an integer
     * @return a boolean, indicate whether matrix contains target
     */
    public boolean searchMatrix(int[][] matrix, int target) {
        // write your code here
        //先二分一下在哪个行，然后再二分一下在哪个列
        if (matrix == null || matrix.length == 0 || matrix[0].length == 0 
        || matrix[0] == null) {
            return false;
        }
        int row = matrix.length;
        int col = matrix[0].length;
        int start = 0;
        int end = row - 1;
        //分一下在哪个行
        while (start + 1 < end) {
            int mid = start + (end - start) / 2;
            if (matrix[mid][0] == target) {
                return true;
            } else if (matrix[mid][0] < target) {
                start = mid;
            } else {
                end = mid;
            }
        }
        //找到行以后细化在哪个列
        if (matrix[end][0] == target || matrix[start][0] == target) {
            return true;
        } else if (matrix[start][0] < target && matrix[end][0] > target) {
            int start1 = 0;
            int end1 = col - 1;
            while (start1 + 1 < end1) {
                int mid1 = start1 + (end1 - start1) / 2;
                if (matrix[start][mid1] == target) {
                    return true;
                } else if (matrix[start][mid1] < target) {
                    start1 = mid1;
                } else {
                    end1 = mid1;
                }
            }
            if (matrix[start][start1] == target) {
                return true;
            } else if (matrix[start][end1] == target) {
                return true;
            } else {
                return false;
            }
        } else {
            int start2 = 0;
            int end2 = col - 1;
            while (start2 + 1 < end2) {
                int mid2 = start2 + (end2 - start2) / 2;
                if (matrix[end][mid2] == target) {
                    return true;
                } else if (matrix[end][mid2] < target) {
                    start2 = mid2;
                } else {
                    end2 = mid2;
                }
            }
            if (matrix[end][start2] == target) {
                return true;
            } else if (matrix[end][end2] == target) {
                return true;
            } else {
                return false;
            }
        }
    }
}

File: leetcode-solution/leetcode74/solution.js
// earch a 2D Matrix
// Write an efficient algorithm that searches for a value in an m x n matrix. This matrix has the following properties:

//     Integers in each row are sorted from left to right.
//     The first integer of each row is greater than the last integer of the previous row.

// Example 1:

// Input:
// matrix = [
//   [1,   3,  5,  7],
//   [10, 11, 16, 20],
//   [23, 30, 34, 50]
// ]
// target = 3
// Output: true
/**
 * param {number[][]} matrix
 * param {number} target
 * return {boolean}
 */

// 思路：
// 1.先用for循环找出在matrix里的哪个数组里
// 2.用二分法找出target是否存在

var searchMatrix = function(matrix, target) {
    if (matrix.length == 0 || matrix[0].length == 0) {
        return false;
    }
    function search(matrix, target) {
        let arr = [];
        let start = 0;
        let end = matrix.length - 1;
        while (start + 1 < end) {
            let mid = Math.floor(start + (end - start) / 2);
            if (matrix[mid] == target) {
                console.log(matrix[mid]);
                return true;
            } else if (matrix[mid] > target) {
                end = mid;
            } else {
                start = mid;
            }
        }
        if (matrix[start] == target) {
            return true;
        }
        if (matrix[end] == target) {
            return true;
        }
        return false;
    }
    for (let i = 0; i <= matrix.length - 1; i++) {
        var aa = matrix[i][0];
        var bb = matrix[i][matrix[i].length - 1];
        if (aa <= target && bb >= target) {
            return search(matrix[i], target);
        }
    }
    return false;
};